common/src/test/java/tachyon/security/authentication/AuthenticationFactoryTest.java

```diff
@@ -22,49 +43,387 @@
+        + "User authentication fails");
+    try {
+      wrongClient.open();
+    } finally {
+      // clean up
+      client.close();
+      mServer.stop();
+    }
+  }
+
+  /**
+   * In CUSTOM mode, if client's username is null, an exception should be thrown in client side.
+   */
+  @Test
+  public void customAuthenticationNullUserTest() throws Exception {
+    mTachyonConf.set(Constants.TACHYON_SECURITY_AUTHENTICATION, "CUSTOM");
+
+    // check case that user is null
+    mThrown.expect(SaslException.class);
+    mThrown.expectMessage("PLAIN: authorization ID and password must be specified");
+    TTransport client = createClient(mTachyonConf, null, "correct-password");
+  }
+
+  /**
+   * In CUSTOM mode, if client's password is null, an exception should be thrown in client side.
+   */
+  @Test
+  public void customAuthenticationNullPasswordTest() throws Exception {
+    mTachyonConf.set(Constants.TACHYON_SECURITY_AUTHENTICATION, "CUSTOM");
+
+    // check case that password is null
+    mThrown.expect(SaslException.class);
+    mThrown.expectMessage("PLAIN: authorization ID and password must be specified");
+    TTransport client = createClient(mTachyonConf, "tachyon", null);
   }
 
-  // TODO: add more tests for methods of AuthenticationFactory
+  /**
+   * In CUSTOM mode, if client's username is empty, an exception should be thrown in server side.
+   */
+  @Test
+  public void customAuthenticationEmptyUserTest() throws Exception {
+    mTachyonConf.set(Constants.TACHYON_SECURITY_AUTHENTICATION, "CUSTOM");
+    mTachyonConf.set(Constants.TACHYON_AUTHENTICATION_PROVIDER_CUSTOM_CLASS,
+        MockAuthenticationProvider.class.getName());
+
+    // start server
+    startServerThread(mTachyonConf);
+
+    // check case that user is empty
+    mThrown.expect(TTransportException.class);
+    mThrown.expectMessage("Peer indicated failure: Plain authentication failed: No authentication"
+        + " identity provided");
+    TTransport client = createClient(mTachyonConf, "", "correct-password");
+    try {
+      client.open();
+    } finally {
+      mServer.stop();
+    }
+  }
+
+  /**
+   * In CUSTOM mode, if client's password is empty, an exception should be thrown in server side.
+   */
+  @Test
+  public void customAuthenticationEmptyPasswordTest() throws Exception {
+    mTachyonConf.set(Constants.TACHYON_SECURITY_AUTHENTICATION, "CUSTOM");
+    mTachyonConf.set(Constants.TACHYON_AUTHENTICATION_PROVIDER_CUSTOM_CLASS,
+        MockAuthenticationProvider.class.getName());
+
+    // start server
+    startServerThread(mTachyonConf);
+
+    // check case that password is empty
+    mThrown.expect(TTransportException.class);
+    mThrown.expectMessage("Peer indicated failure: Plain authentication failed: No password "
+        + "provided");
+    TTransport client = createClient(mTachyonConf, "tachyon", "");
+    try {
+      client.open();
+    } finally {
+      mServer.stop();
+    }
+  }
+
```

Comment:
Done. The case with correct username and password is in `customAuthenticationExactNamePasswordMatchTest()`